File: web/src/shared/connhealth/connhealth.ts
// Mirror of internal/connhealth/connhealth.go — feeds the ConnHealthPill /
// ConnHealthDrawer for web/PWA/mobile clients. The semantics (RTT ring size,
// reconnect window, EMA alpha) MUST match the Go side so both desktop uplink
// and web/mobile show the same numbers.

export const RTT_RING_SIZE = 60
export const RECONNECT_HISTORY_SIZE = 5
export const RECONNECT_WINDOW_MS = 60 * 60 * 1000
const BYTES_EMA_ALPHA = 0.2

export type ConnState = 'closed' | 'connecting' | 'connected' | 'reconnecting'

export interface RTTSample {
  at_ms: number
  rtt_ms: number
}

export interface ReconnectEvent {
  at_ms: number
  reason: string
  duration_ms: number
}

export interface ConnHealthSnapshot {
  state: ConnState
  rtt: {
    last_ms: number | null
    p50_ms: number | null
    p95_ms: number | null
  }
  rtt_samples: RTTSample[]
  reconnect: {
    count_last_hour: number
    last_at_ms: number | null
    last_reason: string
    history: ReconnectEvent[]
  }
  bytes: {
    in_per_sec: number
    out_per_sec: number
  }
  seq_gaps: number
}

type Listener = () => void

export class Tracker {
  private state: ConnState = 'closed'
  private rttRing: RTTSample[]
  private rttHead = 0
  private rttFilled = false

  private reconnects: ReconnectEvent[] = []
  private pendingReason = ''
  private pendingReconnect = false
  private reconnectStartAt = 0

  private bytesInBucket = 0
  private bytesOutBucket = 0
  private bytesBucketSec = 0
  private bytesInPerSec = 0
  private bytesOutPerSec = 0

  private seqGaps = 0

  private listeners: Set<Listener> = new Set()

  constructor() {
    this.rttRing = new Array<RTTSample>(RTT_RING_SIZE)
    for (let i = 0; i < RTT_RING_SIZE; i++) {
      this.rttRing[i] = { at_ms: 0, rtt_ms: 0 }
    }
  }

  onChange(fn: Listener): () => void {
    this.listeners.add(fn)
    return () => { this.listeners.delete(fn) }
  }

  private emit(): void {
    for (const fn of this.listeners) {
      try { fn() } catch { /* swallow */ }
    }
  }

  setState(s: ConnState, nowMS: number): void {
    if (this.state === s) return
    if (s === 'reconnecting') {
      this.pendingReconnect = true
      this.reconnectStartAt = nowMS
    } else if (this.pendingReconnect && s === 'connected') {
      const ev: ReconnectEvent = {
        at_ms: this.reconnectStartAt,
        reason: this.pendingReason,
        duration_ms: nowMS - this.reconnectStartAt,
      }
      this.reconnects.push(ev)
      if (this.reconnects.length > RECONNECT_HISTORY_SIZE) {
        this.reconnects = this.reconnects.slice(-RECONNECT_HISTORY_SIZE)
      }
      this.pendingReconnect = false
      this.pendingReason = ''
    }
    this.state = s
    this.emit()
  }

  setReconnectReason(reason: string, _nowMS: number): void {
    this.pendingReason = reason
  }

  onPongRTT(rttMS: number, nowMS: number): void {
    this.rttRing[this.rttHead] = { at_ms: nowMS, rtt_ms: rttMS }
    this.rttHead = (this.rttHead + 1) % RTT_RING_SIZE
    if (this.rttHead === 0) this.rttFilled = true
  }

  onBytesIn(n: number, nowMS: number): void {
    this.rollover(nowMS)
    this.bytesInBucket += n
  }

  onBytesOut(n: number, nowMS: number): void {
    this.rollover(nowMS)
    this.bytesOutBucket += n
  }

  tick(nowMS: number): void {
    this.rollover(nowMS)
  }

  private rollover(nowMS: number): void {
    const sec = Math.floor(nowMS / 1000)
    if (sec === this.bytesBucketSec) return
    this.bytesInPerSec = BYTES_EMA_ALPHA * this.bytesInBucket + (1 - BYTES_EMA_ALPHA) * this.bytesInPerSec
    this.bytesOutPerSec = BYTES_EMA_ALPHA * this.bytesOutBucket + (1 - BYTES_EMA_ALPHA) * this.bytesOutPerSec
    this.bytesInBucket = 0
    this.bytesOutBucket = 0
    this.bytesBucketSec = sec
  }

  onSeqGap(): void {
    this.seqGaps += 1
  }

  snapshot(nowMS: number): ConnHealthSnapshot {
    const samples = this.orderedRTT()
    const out: ConnHealthSnapshot = {
      state: this.state,
      rtt: { last_ms: null, p50_ms: null, p95_ms: null },
      rtt_samples: samples,
      reconnect: {
        count_last_hour: 0,
        last_at_ms: null,
        last_reason: '',
        history: this.reconnects.slice(),
      },
      bytes: {
        in_per_sec: Math.round(this.bytesInPerSec),
        out_per_sec: Math.round(this.bytesOutPerSec),
      },
      seq_gaps: this.seqGaps,
    }
    if (samples.length > 0) {
      out.rtt.last_ms = samples[samples.length - 1]!.rtt_ms
      const sorted = samples.map(s => s.rtt_ms).sort((a, b) => a - b)
      out.rtt.p50_ms = nearestRank(sorted, 50)
      out.rtt.p95_ms = nearestRank(sorted, 95)
    }
    const cutoff = nowMS - RECONNECT_WINDOW_MS
    out.reconnect.count_last_hour = this.reconnects.filter(e => e.at_ms >= cutoff).length
    if (this.reconnects.length > 0) {
      const last = this.reconnects[this.reconnects.length - 1]!
      out.reconnect.last_at_ms = last.at_ms
      out.reconnect.last_reason = last.reason
    }
    return out
  }

  private orderedRTT(): RTTSample[] {
    if (!this.rttFilled) {
      return this.rttRing.slice(0, this.rttHead)
    }
    return this.rttRing.slice(this.rttHead).concat(this.rttRing.slice(0, this.rttHead))
  }
}

function nearestRank(sorted: number[], p: number): number {
  if (sorted.length === 0) return 0
  if (p <= 0) return sorted[0]!
  if (p >= 100) return sorted[sorted.length - 1]!
  const idx = Math.floor((p * sorted.length) / 100)
  return sorted[Math.min(idx, sorted.length - 1)]!
}

File: internal/connhealth/connhealth.go
// Package connhealth tracks the health of a single WebSocket connection to
// the relay. A Tracker is fed RTT samples, byte counters, state transitions,
// and seq-gap events from the conn's read/write loops, and exposes an
// immutable Snapshot consumable by UI code (via Wails RPC on desktop, via a
// Vue ref on web/mobile).
//
// The Tracker is intentionally pure: no I/O, no goroutines, no timers. All
// time inputs are passed in so tests can drive it deterministically.
package connhealth

import (
	"sort"
	"sync"
	"time"
)

const (
	// RTTRingSize is the number of RTT samples retained. At one sample
	// every 5s, this is a 5-minute sliding window.
	RTTRingSize = 60

	// ReconnectHistorySize is the number of recent reconnect events kept
	// for the drawer's "Recent reconnects" table.
	ReconnectHistorySize = 5

	// ReconnectWindowMS is the rolling window for CountLastHour.
	ReconnectWindowMS = 60 * 60 * 1000

	// bytesEMAAlpha is the smoothing factor for the 1-Hz byte EMAs.
	bytesEMAAlpha = 0.2
)

type State string

const (
	StateClosed       State = "closed"
	StateConnecting   State = "connecting"
	StateConnected    State = "connected"
	StateReconnecting State = "reconnecting"
)

type RTTSample struct {
	AtMS  int64 `json:"at_ms"`
	RTTMS int   `json:"rtt_ms"`
}

type ReconnectEvent struct {
	AtMS       int64  `json:"at_ms"`
	Reason     string `json:"reason"`
	DurationMS int64  `json:"duration_ms"`
}

type RTTStats struct {
	LastMS *int `json:"last_ms"`
	P50MS  *int `json:"p50_ms"`
	P95MS  *int `json:"p95_ms"`
}

type ReconnectStats struct {
	CountLastHour int              `json:"count_last_hour"`
	LastAtMS      *int64           `json:"last_at_ms"`
	LastReason    string           `json:"last_reason"`
	History       []ReconnectEvent `json:"history"`
}

type ByteStats struct {
	InPerSec  int64 `json:"in_per_sec"`
	OutPerSec int64 `json:"out_per_sec"`
}

type Snapshot struct {
	State      State          `json:"state"`
	RTT        RTTStats       `json:"rtt"`
	RTTSamples []RTTSample    `json:"rtt_samples"`
	Reconnect  ReconnectStats `json:"reconnect"`
	Bytes      ByteStats      `json:"bytes"`
	SeqGaps    int            `json:"seq_gaps"`
}

type Tracker struct {
	mu sync.Mutex

	state      State
	stateSince time.Time

	rttRing   [RTTRingSize]RTTSample
	rttCount  int  // total samples ever written (for ring index)
	rttFilled bool // becomes true once rttCount >= RTTRingSize

	reconnects       []ReconnectEvent
	pendingReason    string
	pendingReconnect bool
	reconnectStartAt time.Time

	bytesInBucket  int64
	bytesOutBucket int64
	bytesBucketSec int64
	bytesInPerSec  float64
	bytesOutPerSec float64

	seqGaps int
}

func New() *Tracker {
	return &Tracker{state: StateClosed}
}

// SetState transitions to s at now. The caller is responsible for calling
// SetReconnectReason BEFORE SetState(StateReconnecting) if a reason is known.
func (t *Tracker) SetState(s State, now time.Time) {
	t.mu.Lock()
	defer t.mu.Unlock()
	if t.state == s {
		return
	}
	switch {
	case s == StateReconnecting:
		t.pendingReconnect = true
		t.reconnectStartAt = now
	case t.pendingReconnect && s == StateConnected:
		ev := ReconnectEvent{
			AtMS:       t.reconnectStartAt.UnixMilli(),
			Reason:     t.pendingReason,
			DurationMS: now.Sub(t.reconnectStartAt).Milliseconds(),
		}
		t.reconnects = append(t.reconnects, ev)
		if len(t.reconnects) > ReconnectHistorySize {
			t.reconnects = t.reconnects[len(t.reconnects)-ReconnectHistorySize:]
		}
		t.pendingReconnect = false
		t.pendingReason = ""
	}
	t.state = s
	t.stateSince = now
}

// SetReconnectReason records why the next reconnect happened. Call this from
// the WS close handler (with the mapped reason string) before SetState.
func (t *Tracker) SetReconnectReason(reason string, _ time.Time) {
	t.mu.Lock()
	defer t.mu.Unlock()
	t.pendingReason = reason
}

// OnPongRTT records an RTT sample (ms). Caller should only call this when
// the matching ping's payload round-tripped successfully.
func (t *Tracker) OnPongRTT(rttMS int, now time.Time) {
	t.mu.Lock()
	defer t.mu.Unlock()
	idx := t.rttCount % RTTRingSize
	t.rttRing[idx] = RTTSample{AtMS: now.UnixMilli(), RTTMS: rttMS}
	t.rttCount++
	if t.rttCount >= RTTRingSize {
		t.rttFilled = true
	}
}

// OnBytesIn / OnBytesOut accumulate per-second buckets. The caller MUST call
// Tick(now) once per second (or more frequently — Tick is idempotent within
// the same second).
func (t *Tracker) OnBytesIn(n int, now time.Time)  { t.addBytes(int64(n), 0, now) }
func (t *Tracker) OnBytesOut(n int, now time.Time) { t.addBytes(0, int64(n), now) }

func (t *Tracker) addBytes(in, out int64, now time.Time) {
	t.mu.Lock()
	defer t.mu.Unlock()
	sec := now.Unix()
	if sec != t.bytesBucketSec {
		t.rolloverBuckets(sec)
	}
	t.bytesInBucket += in
	t.bytesOutBucket += out
}

// Tick rolls forward the per-second EMA. Safe to call from a 1 Hz timer.
func (t *Tracker) Tick(now time.Time) {
	t.mu.Lock()
	defer t.mu.Unlock()
	sec := now.Unix()
	if sec != t.bytesBucketSec {
		t.rolloverBuckets(sec)
	}
}

func (t *Tracker) rolloverBuckets(newSec int64) {
	// Apply EMA with the previous bucket's totals, then reset.
	t.bytesInPerSec = bytesEMAAlpha*float64(t.bytesInBucket) + (1-bytesEMAAlpha)*t.bytesInPerSec
	t.bytesOutPerSec = bytesEMAAlpha*float64(t.bytesOutBucket) + (1-bytesEMAAlpha)*t.bytesOutPerSec
	t.bytesInBucket = 0
	t.bytesOutBucket = 0
	t.bytesBucketSec = newSec
}

// OnSeqGap increments the seq-gap counter. The caller observes TypeOut frames
// and increments when the new seq is more than 1 past the last seen.
func (t *Tracker) OnSeqGap() {
	t.mu.Lock()
	defer t.mu.Unlock()
	t.seqGaps++
}

// Snapshot returns an immutable view at the given wall-clock now.
func (t *Tracker) Snapshot(now time.Time) Snapshot {
	t.mu.Lock()
	defer t.mu.Unlock()

	s := Snapshot{
		State:      t.state,
		RTTSamples: []RTTSample{},
		Reconnect:  ReconnectStats{History: []ReconnectEvent{}},
	}
	s.SeqGaps = t.seqGaps
	s.Bytes.InPerSec = int64(t.bytesInPerSec)
	s.Bytes.OutPerSec = int64(t.bytesOutPerSec)

	samples := t.orderedRTTLocked()
	if len(samples) > 0 {
		s.RTTSamples = samples
		last := samples[len(samples)-1].RTTMS
		s.RTT.LastMS = &last
		sorted := make([]int, len(samples))
		for i, sm := range samples {
			sorted[i] = sm.RTTMS
		}
		sort.Ints(sorted)
		p50 := percentile(sorted, 50)
		p95 := percentile(sorted, 95)
		s.RTT.P50MS = &p50
		s.RTT.P95MS = &p95
	}

	nowMS := now.UnixMilli()
	cutoff := nowMS - ReconnectWindowMS
	count := 0
	for _, ev := range t.reconnects {
		if ev.AtMS >= cutoff {
			count++
		}
	}
	s.Reconnect.CountLastHour = count
	if len(t.reconnects) > 0 {
		s.Reconnect.History = append([]ReconnectEvent(nil), t.reconnects...)
		last := t.reconnects[len(t.reconnects)-1]
		s.Reconnect.LastAtMS = &last.AtMS
		s.Reconnect.LastReason = last.Reason
	}
	return s
}

func (t *Tracker) orderedRTTLocked() []RTTSample {
	if !t.rttFilled {
		return append([]RTTSample(nil), t.rttRing[:t.rttCount]...)
	}
	start := t.rttCount % RTTRingSize
	out := make([]RTTSample, 0, RTTRingSize)
	out = append(out, t.rttRing[start:]...)
	out = append(out, t.rttRing[:start]...)
	return out
}

func percentile(sorted []int, p int) int {
	if len(sorted) == 0 {
		return 0
	}
	if p <= 0 {
		return sorted[0]
	}
	if p >= 100 {
		return sorted[len(sorted)-1]
	}
	// Nearest-rank.
	idx := (p * len(sorted)) / 100
	if idx >= len(sorted) {
		idx = len(sorted) - 1
	}
	return sorted[idx]
}
